libraries/Microsoft.Bot.Builder.Ai/TranslationMiddleware.cs

```diff
@@ -0,0 +1,177 @@
+        /// <param name="translatorKey"></param>
+        /// <param name="patterns">Dictionary with language as a key and list of patterns as value</param>
+        /// <param name="getUserLanguage">Delegate for getting the user language</param>
+        /// <param name="setUserLanguage">Delegate for setting the user language, returns true if the language was changed (implements logic to change language by intercepting the message)</param>
+        public TranslationMiddleware(string[] nativeLanguages, string translatorKey, Dictionary<string, List<string>> patterns, Func<ITurnContext, string> getUserLanguage, Func<ITurnContext, Task<bool>> setUserLanguage)
+        {
+            if (nativeLanguages == null || nativeLanguages.Count() == 0)
+                throw new ArgumentNullException(nameof(nativeLanguages));
+            this._nativeLanguages = nativeLanguages;
+            if (string.IsNullOrEmpty(translatorKey))
+                throw new ArgumentNullException(nameof(translatorKey));
+            this._translator = new Translator(translatorKey);
+            this._patterns = patterns ?? throw new ArgumentNullException(nameof(patterns));
+            this._patterns = patterns;
+            this._getUserLanguage = getUserLanguage ?? throw new ArgumentNullException(nameof(getUserLanguage));
+            this._setUserLanguage = setUserLanguage ?? throw new ArgumentNullException(nameof(setUserLanguage)); 
+        }
+
+        /// <summary>
+        /// Incoming activity
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="next"></param>
+        /// <returns></returns>       
+        public async Task OnProcessRequest(ITurnContext context, MiddlewareSet.NextDelegate next)
+        {
+            IMessageActivity message = context.Activity.AsMessageActivity();
+            if (message != null)
+            {
+                if (!String.IsNullOrWhiteSpace(message.Text))
+                {
+
+                    var languageChanged = false;
+
+                    if (_setUserLanguage != null)
+                    {
+                        languageChanged = await _setUserLanguage(context);
+                    }
+
+                    if (!languageChanged)
+                    {
+                        // determine the language we are using for this conversation
+                        var sourceLanguage = "";
+                        if (_getUserLanguage == null)
+                            sourceLanguage = await Task.Run(() => _translator.Detect(message.Text)); // context.Conversation.Data["Language"]?.ToString() ?? this._nativeLanguages.FirstOrDefault() ?? "en";
+                        else
+                        {
+                            sourceLanguage = _getUserLanguage(context);
+                        } 
+                        if (_patterns.ContainsKey(sourceLanguage) && _patterns[sourceLanguage].Count>0)
```

Comment:
Try as I might, I can't figure out what patterns are doing here. Could you include some comments?